### PR TITLE
Validate template parameter name and description length

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

File: pkg/template/api/validation/validation.go [hunk 1]
```diff
@@ -12,15 +12,28 @@ import (
 
 var parameterNameExp = regexp.MustCompile(`^[a-zA-Z0-9\_]+$`)
 
+const (
+	// FIXME: There should be global constant specifying the maximum allowed
+	//				length for resources.
+	maxParameterNameLength        = 120
```
Comment (smarterclayton) [inline pkg/template/api/validation/validation.go]: Why is this so short?

File: pkg/template/api/validation/validation.go [hunk 2]
```diff
@@ -12,15 +12,28 @@ import (
...
+	maxParameterDescriptionLength = 255
```
Comment (smarterclayton) [inline pkg/template/api/validation/validation.go]: Too short

**Comment by mfojtik** [comment]
@liggitt @danmcp PTAL

File: pkg/template/api/validation/validation.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/template/api/validation/validation.go]: You should validate parameters like annotations - total size matters more than individual length (name length is reasonable, everything else is arbitrary).

Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: @smarterclayton suggestions? can I take some constant from upstream if such thing exists?

Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: @smarterclayton https://github.com/openshift/origin/issues/2689#issuecomment-107957167

Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: @smarterclayton so total size of all descriptions <64kb ?

File: pkg/template/api/validation/validation.go (hunk 1 above)
Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: 255? note that this is parameter name, not the actual env var

**Comment by mfojtik** [comment]
@smarterclayton updated, is it better now? (in v2 we limit size of env var name to 128 chars)

File: pkg/template/api/validation/validation.go [hunk 3]
```diff
@@ -33,6 +40,11 @@ func ValidateProcessedTemplate(template *api.Template) fielderrors.ValidationErr
 func ValidateTemplate(template *api.Template) (errs fielderrors.ValidationErrorList) {
 	errs = validation.ValidateObjectMeta(&template.ObjectMeta, true, validation.ValidatePodName).Prefix("metadata")
 	errs = append(errs, validateTemplateBody(template)...)
+	fields := map[string]string{}
+	for i := range template.Parameters {
+		fields[template.Parameters[i].Name] = template.Parameters[i].Description
+	}
+	errs = append(errs, validation.ValidateAnnotations(fields, "description")...)
```
Comment (liggitt) [inline pkg/template/api/validation/validation.go]: `ValidateAnnotations` checks that the keys are qualified names... if we're already checking `if !parameterNameExp.MatchString(param.Name) {`, will we get double errors on bad key names?

Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: @liggitt I think we can defer the validation of keys to k8s and remove our regexp

Comment (liggitt) [inline pkg/template/api/validation/validation.go]: as long as the substitution code handles all the values the annotation key validation allows, that's fine

Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: it does not... and the key length is restricted to 63 character which will make @smarterclayton sad... I think I will have to copy the func from k8s and validate keys based on our regexp...

**Comment by mfojtik** [comment]
@smarterclayton @liggitt fixed, we now check totalSize < 64k for 'name', 'description', 'value' and 'from' fields... If you think we should have more (this is default for k8s annotations) tell me, since we have 2 more fields there, I'm OK bumping it to 128k

**Comment by mfojtik** [comment]
[test]

File: pkg/template/api/validation/validation.go [hunk 4]
```diff
@@ -12,6 +12,26 @@ import (
 
 var parameterNameExp = regexp.MustCompile(`^[a-zA-Z0-9\_]+$`)
 
+// totalParametersSizeLimitB is the total size limit for the template
+// parameters names, values, froms and descriptions
+// TODO: This constant was copied from k8s totalAnnotationSizeLimitB, we
+//       should make the upstream public with a better name and re-use it
+const totalParametersSizeLimitB int = 64 * (1 << 10) // 64 kB
```
Comment (liggitt) [inline pkg/template/api/validation/validation.go]: `64 * 1024` is more inspectable

Comment (mfojtik) [inline pkg/template/api/validation/validation.go]: @liggitt agree

**Comment by mfojtik** [comment]
@liggitt updated, PTAL

File: pkg/template/api/validation/validation.go [hunk 5]
```diff
@@ -12,6 +12,26 @@ import (
...
+//       should make the upstream public with a better name and re-use it
+const totalParametersSizeLimitB int = 64 * 1024 // 64 kB
+
+// ValidateParametersSize validates the total size of all parameters names,
+// descriptions.
+func ValidateParametersSize(params []api.Parameter) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+	var totalSize int64
+	for i := range params {
+		totalSize += (int64)(len(params[i].Name)) + (int64)(len(params[i].Description))
```
Comment (liggitt) [inline pkg/template/api/validation/validation.go]: I thought we said we cared about value length when the template was going to be stored, and we didn't care when it was just going to be processed. @smarterclayton, what was the reason for validating the total size again?

Comment (smarterclayton) [inline pkg/template/api/validation/validation.go]: Storage size is all we care about.  We should have a separate validation path for process (i believe) that should not be so restrictive.

----- Original Message -----

**Comment by danmcp** [comment]
@mfojtik Needs rebase and is this for 3.0?

**Comment by mfojtik** [comment]
@danmcp this is post-3.0, but needs rebase anyway :)

**Comment by smarterclayton** [comment]
Reopen when this is ready (and takes into account the comments).

**Pull request closed without merging**